src/optimized_transaction.rs

```diff
@@ -0,0 +1,210 @@
+    /// should be skipped, and how many times to retry the transaction, if provided
+    ///
+    /// # Returns
+    /// The transaction signature, if successful
+    pub async fn send_smart_transaction(&self, config: SmartTransactionConfig<'_>) -> Result<Signature> {
+        let pubkey: Pubkey = config.from_keypair.pubkey();
+        let mut recent_blockhash: Hash = self.connection().get_latest_blockhash()?;
+
+        // Build the initial transaction and estimate the priority fee
+        let mut transaction: Transaction = Transaction::new_with_payer(&config.instructions, Some(&pubkey));
+        transaction.try_sign(&[config.from_keypair], recent_blockhash)?;
+
+        // Serialize the transaction
+        let serialized_transaction: Vec<u8> =
+            serialize(&transaction).map_err(|e: Box<ErrorKind>| HeliusError::InvalidInput(e.to_string()))?;
+
+        // Convert the serialized transaction to a Base64 string
+        let transaction_base64: String = STANDARD.encode(&serialized_transaction);
+
+        // Get the priority fee estimate based on the serialized transaction
+        let priority_fee_request: GetPriorityFeeEstimateRequest = GetPriorityFeeEstimateRequest {
+            transaction: Some(transaction_base64),
+            account_keys: None,
+            options: None,
+        };
+
+        let priority_fee_estimate: GetPriorityFeeEstimateResponse =
+            self.rpc().get_priority_fee_estimate(priority_fee_request).await?;
+
+        let priority_fee = priority_fee_estimate
+            .priority_fee_estimate
+            .ok_or(HeliusError::InvalidInput(
+                "Priority fee estimate not available".to_string(),
+            ))?
+            .to_bits();
+
+        // Add the compute unit price instruction with the estimated fee
+        let compute_budget_ix: Instruction = ComputeBudgetInstruction::set_compute_unit_price(priority_fee);
+        let mut final_instructions: Vec<Instruction> = vec![compute_budget_ix];
+        final_instructions.extend(config.instructions.clone());
+
+        // Get the optimal compute units
+        if let Some(units) = self
+            .get_compute_units(final_instructions.clone(), pubkey, vec![])
+            .await?
+        {
+            // Add some margin to the compute units
+            let compute_units_ix: Instruction =
+                ComputeBudgetInstruction::set_compute_unit_limit((units as f64 * 1.1).ceil() as u32);
+            final_instructions.insert(0, compute_units_ix);
+        }
+
+        // Build the optimized transaction
+        let mut optimized_transaction: Transaction = Transaction::new_with_payer(&final_instructions, Some(&pubkey));
+        optimized_transaction.try_sign(&[config.from_keypair], recent_blockhash)?;
+
+        // Re-fetch the blockhash every 4 retries, or roughly once every minute
+        let blockhash_validity_threshold: usize = 4;
```

Comment:
this should ideally be set based on time. If you change retry interval it will make this go out of whack